impl/spanner/directory/directory.go

```diff
@@ -0,0 +1,221 @@
+	}, nil
+}
+
+// List returns all Directories. showDeleted indicates whether deleted directories should be returned.
+func (t *Table) List(ctx context.Context, showDeleted bool) ([]*directory.Directory, error) {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+
+	ret := []*directory.Directory{}
+	err = rtx.Read(ctx, table, spanner.AllKeys(), directoriesCols).Do(
+		func(r *spanner.Row) error {
+			d, err := readRow(r)
+			if err != nil {
+				return err
+			}
+			if d.Deleted && !showDeleted {
+				return nil
+			}
+			ret = append(ret, d)
+			return nil
+		})
+	return ret, err
+}
+
+// Write creates a new Directory.
+func (t *Table) Write(ctx context.Context, dir *directory.Directory) error {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return err
+	}
+	// Prepare data.
+	anyPB, err := ptypes.MarshalAny(dir.VRFPriv)
+	if err != nil {
+		return err
+	}
+	anyData, err := proto.Marshal(anyPB)
+	if err != nil {
+		return err
+	}
+
+	tmap, err := proto.Marshal(dir.Map)
+	if err != nil {
+		return err
+	}
+	tlog, err := proto.Marshal(dir.Log)
+	if err != nil {
+		return err
+	}
+
+	m, err := spanner.InsertStruct(table, spannerDirectory{
+		DirectoryID:   dir.DirectoryID,
+		Map:           tmap,
+		Log:           tlog,
+		VRFPublicKey:  dir.VRF.GetDer(),
+		VRFPrivateKey: anyData,
+		MinInterval:   dir.MinInterval.Nanoseconds(),
+		MaxInterval:   dir.MaxInterval.Nanoseconds(),
+		Deleted:       false,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Apply(ctx, []*spanner.Mutation{m})
+	return err
+}
+
+// Read retriaves a directory from storage. Returns status.NotFound if the row is deleted.
+func (t *Table) Read(ctx context.Context, directoryID string, showDeleted bool) (*directory.Directory, error) {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+	row, err := rtx.ReadRow(ctx, table, spanner.Key{directoryID}, directoriesCols)
+	if err != nil {
+		return nil, err
+	}
+	dir, err := readRow(row)
+	if err != nil {
+		return nil, err
+	}
+	if !showDeleted && dir.Deleted {
+		return nil, status.Errorf(codes.NotFound, "Directory %v is deleted", directoryID)
+	}
+	return dir, nil
+}
+
+// unwrapAnyProto returns the proto object seralized inside a serialized any.Any
+func unwrapAnyProto(anyData []byte) (proto.Message, error) {
+	var anyPB any.Any
+	if err := proto.Unmarshal(anyData, &anyPB); err != nil {
+		return nil, err
+	}
+	var privKey ptypes.DynamicAny
+	if err := ptypes.UnmarshalAny(&anyPB, &privKey); err != nil {
+		return nil, err
+	}
+	return privKey.Message, nil
+}
+
+// SetDelete deletes or undeletes a directory.
+func (t *Table) SetDelete(ctx context.Context, directoryID string, isDeleted bool) error {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return err
+	}
+	_, err = client.Apply(ctx, []*spanner.Mutation{
+		spanner.Update(table, []string{"DirectoryID", "Deleted", "DeleteTime"},
```

Comment:
Can `DeleteTime` be auto-filled with Spanner's commit timestamp instead of the user-space time? There is `spanner.CommitTimestamp` const for this purpose. Warning: Might require tweaking the schema a bit, so that the field has the right metadata allowing to use it for commit timestamps.

Reply:
It probably could be, but I don't see a strong reason to do this.